middleware/http/opa/opa_middleware.go

```diff
@@ -0,0 +1,178 @@
+package opa
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"strings"
+
+	"github.com/dapr/components-contrib/middleware"
+	"github.com/dapr/dapr/pkg/logger"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/valyala/fasthttp"
+)
+
+type opaMiddlewareMetadata struct {
+	Rego            string `json:"rego"`
+	IncludedHeaders string `json:"includedHeaders"`
+}
+
+// NewOpaMiddleware returns a new Open Policy Agent middleware
+func NewOpaMiddleware(logger logger.Logger) *Middleware {
+	return &Middleware{logger: logger}
+}
+
+// Middleware is an OPA  middleware
+type Middleware struct {
+	logger logger.Logger
+}
+
+// RegoResult is the expected result from rego policy
+type RegoResult struct {
+	Allow             bool              `json:"allow"`
+	AdditionalHeaders map[string]string `json:"additional_headers,omitempty"`
+	StatusCode        int               `json:"status_code,omitempty"`
+}
+
+// GetHandler retruns the HTTP handler provided by the middleware
+func (m *Middleware) GetHandler(metadata middleware.Metadata) (func(h fasthttp.RequestHandler) fasthttp.RequestHandler, error) {
+	meta, err := m.getNativeMetadata(metadata)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+
+	query, err := rego.New(
+		rego.Query("result = data.http.allow"),
+		rego.Module("inline.rego", meta.Rego),
+	).PrepareForEval(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return func(h fasthttp.RequestHandler) fasthttp.RequestHandler {
+		return func(ctx *fasthttp.RequestCtx) {
+			if handled := m.evalRequest(ctx, meta, &query); handled {
+				return
+			}
+			h(ctx)
+		}
+	}, nil
+}
+
+func (m *Middleware) evalRequest(ctx *fasthttp.RequestCtx, meta *opaMiddlewareMetadata, query *rego.PreparedEvalQuery) bool {
+	headers := map[string]string{}
+	var allowedHeaders = strings.Split(meta.IncludedHeaders, ",")
+	ctx.Request.Header.VisitAll(func(key, value []byte) {
+		for _, allowed := range allowedHeaders {
+			if string(key) == allowed {
+				headers[string(key)] = string(value)
+			}
+		}
+	})
+
+	queryArgs := map[string][]string{}
+	ctx.QueryArgs().VisitAll(func(key, value []byte) {
+		if val, ok := queryArgs[string(key)]; ok {
+			queryArgs[string(key)] = append(val, string(value))
+		} else {
+			queryArgs[string(key)] = []string{string(value)}
+		}
+	})
+
+	input := map[string]interface{}{
+		"request": map[string]interface{}{
+			"method":    string(ctx.Method()),
+			"path":      string(ctx.Path()),
+			"raw_query": string(ctx.QueryArgs().QueryString()),
+			"query":     queryArgs,
+			"headers":   headers,
+			"scheme":    string(ctx.Request.URI().Scheme()),
+			//TODO: allow opting into body support? Reading body isn't efficient
```

Comment:
Please file tracking issues for these TODO as these are very hard to keep track and follow up later.

After that, we can replace `TODO` with `TODO(#issue)` or `TODO(<link-to-issue-comment>)`.

Reply:
Wanted to make sure the `opa` middleware was a go before I extracted these out. If consensus is yes on adding the middleware, I can get these marshaled to issues.

Reply:
This was discussed in a maintainers meeting and was decided to be included, provided the PR is approved and that there's a docs update for the middleware :)

Reply:
@yaron2 Cool beans! I'll spin out these tickets and start a docs pr then